ndb/src/google/cloud/ndb/tasklets.py

```diff
@@ -44,18 +49,77 @@ def add_flow_exception(*args, **kwargs):
 
 
 class Future:
-    __slots__ = ()
+    """Base class for all NDB futures.
 
-    def __init__(self, *args, **kwargs):
-        raise NotImplementedError
+    Provides interface defined by :class:`concurrent.futures.Future` as well as
+    legacy NDB ``Future`` class.
+
+    Sketchy code, very incomplete.
+    """
+    def __init__(self):
+        self._done = False
+        self._result = None
+
+    def set_result(self, result):
+        self._result = result
+        self._done = True
+
+    def done(self):
+        return self._done
+
+    def result(self):
+        while not self._done:
+            _eventloop.run1()
+        return self._result
+
+
+class TaskletFuture(Future):
+    """A Future for an NDB tasklet.
+
+    Sketchy code, very incomplete.
+    """
+    def __init__(self, generator):
+        super(TaskletFuture, self).__init__()
+        self.generator = generator
+
+    def _advance_tasklet(self, send_value=None):
+        """Advances tasklet one step given a value to send into generator."""
+        try:
+            # Send next value in to generator (will be None if just starting)
+            yielded = self.generator.send(send_value)
+
+        except StopIteration as stop:
+            # Tasklet has finished
+            self.set_result(get_return_value(stop))
+
+        else:
```

Comment:
Why is the `else:` needed rather than just continuing on?

Reply:
If you're asking why it isn't in the 'try' block, probably to make it more obvious, visually, where the exception is expected to get thrown. Why else would there be an 'else' for 'try/except' in the language? FWIW, I copied that structure from the legacy code and didn't think very hard about it as it seemed intuitive.

If you're asking why it isn't just after the 'try/except', because that code doesn't get run if the tasklet has finished.

Reply:
> If you're asking why it isn't just after the 'try/except', because that code doesn't get run if the tasklet has finished.

I'm not grokking this, can you explain a bit more?

Reply:
When we send a value into the tasklet generator, one of two things will happen: the generator will yield something, or it will raise StopIteration.

The code in the "else" block handles the case where the tasklet generator has yielded something, where "something" is always another tasklet's Future, a sequence of Futures (parallel yield), or a Future from gRPC. The tasklet still has work to do and this block coordinates moving that work along.

The code in the "except" clause handles the StopIteration case, which is how we know the tasklet has finished. The code in this block sets the Future's result and the Future has now finished all of its work.

There would never be a case where where we want to run the code in both clauses in the same call.

Reply:
Right, but you could `return` from the `except StopIteration:` block. I was just advocating for a dedent (that comes with ditching the `else:`). The more indents, the harder it is to grok code, that's all.

Reply:
Maybe. One could also argue that seeing the three blocks visually, makes it easier to see what's going on, as opposed to having to notice that the StopIteration block has a return. I can see arguments for both.

Reply:
It's your call. I've always seen "exit early, fewer indents", e.g. http://blog.timoxley.com/post/47041269194/avoid-else-return-early. No need to keep going round on this though.

Reply:
Yeah, it's a judgment call. The examples in that blog post are clearly better refactored.